engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/internal/PairwiseFloatRingBuffer.java

```diff
@@ -87,72 +97,139 @@ private void evaluateRangeFast(int start, int end) {
+        // overlap and the calculation simplifies).
+
+        final float value;
+
+        if (pushDirty && popDirty) {
+            if (dirtyPushHead > dirtyPushTail && dirtyPopHead > dirtyPopTail) {
+                // both are wrapped
+                value = evaluateTree(capacity, Math.max(dirtyPushTail, dirtyPopTail), Math.min(dirtyPushHead, dirtyPopHead), chunkSize - 1);
+            } else if (dirtyPushHead > dirtyPushTail) {
+                // push wrapped, pop is not
+                value = evaluateTree(capacity, dirtyPushTail, dirtyPopHead, dirtyPopTail, dirtyPushHead, chunkSize - 1);
+            } else if (dirtyPopHead > dirtyPopTail) {
+                // pop wrapped, push is not
+                value = evaluateTree(capacity, dirtyPopTail, dirtyPushHead, dirtyPushTail, dirtyPopHead, chunkSize - 1);
+            } else {
+                // neither wrapped
+                if (dirtyPushHead > dirtyPopHead) {
+                    value = evaluateTree(dirtyPopHead, dirtyPopTail, dirtyPushHead, dirtyPushTail);
+                } else {
+                    value = evaluateTree(dirtyPushHead, dirtyPushTail, dirtyPopHead, dirtyPopTail);
                 }
             }
-            // how far should we advance
-            final int nextIndex = dirtyIndex + 1;
-            if (nextIndex < dirtyIndices.size() && dirtyIndices.get(nextIndex) == right) {
-                dirtyIndex += 2;
+        } else if (pushDirty) {
+            if (dirtyPushHead > dirtyPushTail) {
+                value = evaluateTree(capacity, dirtyPushTail, dirtyPushHead, chunkSize - 1);
             } else {
-                dirtyIndex++;
+                value = evaluateTree(dirtyPushHead, dirtyPushTail);
             }
+        } else if (popDirty) {
+            if (dirtyPopHead > dirtyPopTail) {
+                value = evaluateTree(capacity, dirtyPopTail, dirtyPopHead, chunkSize - 1);
+            } else {
+                value = evaluateTree(dirtyPopHead, dirtyPopTail);
+            }
+        } else {
+            value = storageChunk.get(1);
         }
-        allDirty = false;
-        dirtyIndices.clear();
 
-        // final value is in index 1
-        return storageChunk.get(1);
+        clearDirty();
+
+        return value;
     }
 
     private void grow(int increase) {
         int oldCapacity = capacity;
         int oldChunkSize = chunkSize;
 
-        int size = size();
+        // assert that we are not asking for the impossible
+        Assert.eqTrue(PAIRWISE_MAX_CAPACITY - increase >= size, "PairwiseFloatRingBuffer size <= PAIRWISE_MAX_CAPACITY");
 
         final int minLength = size + increase;
 
         // double the current capacity until there is sufficient space for the increase
         while (capacity <= minLength) {
             capacity *= 2;
-            chunkSize = capacity * 2;
+            chunkSize = Math.min(capacity * 2, PAIRWISE_MAX_CAPACITY);
```

Comment:
I think you can short circuit, with your assert, to the power of two that encloses minLength.